client/src/connection/rest/job.ts

```diff
@@ -0,0 +1,175 @@
+
+    return await this.requestLink(link, options);
+  }
+
+  async getState(options?: stateOptions): Promise<string> {
+    const parms: JobsApiGetJobStateRequest = {
+      sessionId: this._self.sessionId,
+      jobId: this.id,
+      wait: options?.wait,
+      ifNoneMatch: options?.onChange ? this.etag : undefined,
+    };
+
+    const resp = await this.api.getJobState(parms);
+    if (resp.status === 200) {
+      //Set the new etag
+      this.etag = resp.etag;
+
+      //return the state
+      return resp.data;
+    } else if (resp.status === 304) {
+      //Not modified
+      return await this.getState(); //This is bad. We need to cache the last state value
+    } else {
+      throw new Error("Something went wrong");
+    }
+  }
+
+  //Get the job log as a stream.
+  async *getLogStream(options?: { timeout?: number }): AsyncGenerator<LogLine> {
+    let state = await this.getState();
+    const timeout = options?.timeout ?? 10;
+    let start = 0;
+
+    const states = ["done", "canceled", "error", "warning", "completed"];
+
+    while (states.indexOf(state) === -1) {
+      //Get a log page
+      const resp = await this.logs.getJobLog({
+        sessionId: this._self.sessionId,
+        jobId: this.id,
+        start: start,
+        timeout: timeout,
+      });
+
+      if (resp.status === 200) {
+        const items = resp.data.items;
+        const num = items.length;
+        for (let i = 0; i < num; i++) {
+          yield items[i];
+        }
+
+        //increase start location
+        start += num;
+
+        //get new state
+        state = await this.getState();
+      } else {
+        break;
+      }
+    }
+
+    //There is a chance that the job ended between our last read and now.
+    //Need to make sure we clear out the log
+
+    let nextLink: Link = undefined;
+    let resp = await this.logs.getJobLog({
+      sessionId: this._self.sessionId,
+      jobId: this.id,
+      start: start,
+      timeout: timeout,
+    });
+
+    //To clear out the log, we yeild all lines until there is not "next" link
+    do {
+      if (resp.status === 200) {
+        nextLink = resp.links?.find((link) => link.rel === "next");
+        const items = resp.data.items;
+        const num = items.length;
+        for (let i = 0; i < num; i++) {
+          yield items[i];
+        }
+
+        if (nextLink) {
+          resp = await this.requestLink(nextLink);
+        }
+      } else {
+        break;
+      }
+    } while (nextLink !== undefined);
+  }
+
+  /*
+  Check to see if the job is done.
+  Done is defined as the job having run to some sort of completed state
+  ie. the job stopped because it finished executing or is was stopped 
+  due to an error or other signal.
+  */
+  async isDone(state?: string): Promise<boolean> {
+    const doneStates = ["done", "canceled", "error", "warning", "completed"];
+
+    if (state === undefined) {
+      state = await this.getState();
+    }
+
+    return doneStates.indexOf(state) === -1;
```

Comment:
Instead of reassigning a param, could we...
```js
return doneStates.indexOf(state || await this.getState()) === -1;
```

Reply:
> Instead of reassigning a param, could we...
> 
> ```js
> return doneStates.indexOf(state || await this.getState()) === -1;
> ```

yeah, I could do that. Still getting used to ts/js shortcuts.

The getState function can never lead to recursion. The only way a 304 is returned is if the if-none-match header is sent. calling getState with no options, will send a simple call to getState which simply returns the current state of the job.